### ISSUE TITLE
Training and sampling speed

### ISSUE THREAD

**Issue by amasawa** [yuanzhi-zhu/SlimFlow#3]
Thank you for sharing the code. I just wondered what the sampling and training speeds are, because the ncsnpp cost too much time for sampling and training.

**Comment by yuanzhi-zhu** [comment]
SlimFlow aims to train a smaller and faster flow (from scratch using teacher generated samples).
The training may take days on 4090, the sampling only takes 1 step or few steps.

**Comment by amasawa** [comment]
Thanks for your quick response! I was wondering if the training speed is reasonable. I followed the description in your paper and attempted to train the model from scratch using the following command:
`python ./train.py \
    --config ./configs/rectified_flow/cifar10_rf_gaussian.py  \
    --config.expr 1_rectified_flow \
`
The training setup uses a batch size of 128 and iterations = 1300001. I'm running this on 2 NVIDIA H100 GPUs, and the estimated training time is **28 days**. Does this align with your experiment?

**Comment by yuanzhi-zhu** [comment]
Hey, the command line you provide is meant to train a 1 flow.
In this work, you donot necessary to do that. You can adopt any existing diffusion/flow model to generate paired sample and run slimflow to train a smaller model. 5 days is more than enough to get a good model in my case.
But SlimFlow is in general slow in training, comparing to other method such as https://github.com/locuslab/ect

**Comment by amasawa** [comment]
Thank you for your detailed explanation in the issue and README.

I have a few questions regarding the training process for SLIM. From my understanding:

1. We need to generate data pairs from the teacher using the following code:  
   `python ./generate_data.py \
    --config ./configs/rectified_flow/cifar10_rf_gaussian.py  \
    --ckpt_path "logs/1_rectified_flow/checkpoints/checkpoint_14.pth" \
    --data_root "reflow_data/1_rectified_flow_50000/" \
    --config.sampling.total_number_of_samples 50000 --config.seed 0 \
    --config.training.batch_size 512 \
    --config.sampling.direction from_z0 \`  
   (By the way, could you clarify the differences in 3 options: 'from_z0', 'from_z1', 'random_paired'?)

2. After generating the data pairs, we train the smaller student model (Annealing Reflow) using these pairs (with the \(\beta\) selection mechanism automatically implemented during training) using the following code:  
   `python ./train.py \
    --config ./configs/rectified_flow/cifar10_rf_gaussian.py  \
    --config.expr 2_rectified_flow_500001flow_flip_warmup_300000_28m \
    --config.flow.h_flip=true \
    --config.training.x0_randomness warmup_300000 \
    --config.training.snapshot_freq 50000 \
    --config.training.snapshot_sampling 10000 \
    --config.data.reflow_data_root "reflow_data/1_rectified_flow_50000/" \
    --config.model.nf 128 --config.model.num_res_blocks 2 \
    --config.model.ch_mult '(1, 2, 2)' \`

3. Finally, we train the distilled model on the data pairs using the following code:  
   `python ./train.py \
    --config ./configs/rectified_flow/cifar10_rf_gaussian.py  \
    --config.data.reflow_data_root "reflow_data/1_rectified_flow_50000/" \
    --config.flow.flow_t_schedule t0 \
    --config.training.loss_type lpips \
    --config.flow.use_teacher true \
    --config.expr 2_rectified_flow_500000bigflow_28m_distill_lpips_use_teacher \
    --config.flow.pre_train_model "./logs/2_rectified_flow_500001flow_flip_warmup_300000_28m/checkpoints/checkpoint_16.pth" \
    --config.model.nf 128 --config.model.num_res_blocks 2 \
    --config.model.ch_mult '(1, 2, 2)' \`

**Comment by yuanzhi-zhu** [comment]
Hey @amasawa , your understanding is mostly correct! Note that we usually re-generate data pairs using annealed flows for distillation, so there is a step 2.5 used to generate synthetic data.
when creating paired data with flow models, we can simulate from noise to data, which is the option 'from_z0'. Alternatively, we can also simulate the ODE from data to noise (similar to DDIM inversion in diffusion models) to create data pairs with real data in the training set as z1. (However, due to the discrete error and etc., 'from_z1' doesnot lead to better 2flow, at least according to our experiments. 'random_paired' with return 50000 (z0,z1) pairs, with z1 from training set and z0 iid samples from initial distribution. We use this for comparison to well paired data.
please donot hesitate if you have any further questions

**Comment by amasawa** [comment]
Thanks, Yuanzhi! Your explanation was very clear and helpful!

I still have some questions regarding Figure 4. I noticed that Reflow does not provide a pretrained model across different model sizes.  

1. Should we train the 2-rectified flow with different solvers (RK45 for the blue line and Euler for the red line) guided by the 1-rectified flow, using various model sizes (3M, 7M, 15M, and 28M)?  
2. Were these experiments designed to evaluate the performance of 2-rectified flow under different solvers?  
3. Additionally, could you explain what SlimFlow represents for the green line?

**Comment by yuanzhi-zhu** [comment]
The only input to the SlimFlow Algorithm is the data pairs generated by the teacher model, which can be either EDM or 1-flow in our case.
1. We use the best (to our knowledge) sampler for the teacher model to generate the data pairs. For EDM is the heun sampler, and for 1-flow is the rk45 sampler, regardless of the model sizes.
2. Yes, and to show the gap between distilled one-step model and the 2-flows
3. It refers to the final distilled one-step model
